### PR TITLE
Use single file upgrades where possible.

### PR DESCRIPTION
**Pull request by alexet**
Where possible use the new single file form of upgrades. Older versions of `codeql` simply ignore the `singleFileUpgrades` property and give multifile upgrades. Hence we can just set the new flag and see which sort of upgrade we get pack. If a version of codeql returns a multi file upgrade then it certainly supports running them so there are no issues there.

### CODE REVIEW

File: extensions/ql-vscode/src/upgrades.ts [hunk 1]
```diff
@@ -150,7 +150,11 @@ export async function upgradeDatabase(
 
   try {
     qs.logger.log('Running the following database upgrade:');
-    qs.logger.log(compileUpgradeResult.compiledUpgrades.scripts.map(s => s.description.description).join('\n'));
+    if (compileUpgradeResult.compiledUpgrades.compiledUpgradeFile === undefined) {
+      qs.logger.log(compileUpgradeResult.compiledUpgrades.scripts.map(s => s.description.description).join('\n'));
+    } else {
+      qs.logger.log(compileUpgradeResult.compiledUpgrades.descriptions.map(s => s.description).join('\n'));
+    }
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/upgrades.ts:157]: nit: Can you restructure the if expression? I find it hard to reason about. Something like this I find easier to read

```suggestion
    if (compileUpgradeResult.compiledUpgrades.compiledUpgradeFile) {
      // Single file upgrade
      qs.logger.log(compileUpgradeResult.compiledUpgrades.descriptions.map(s => s.description).join('\n'));
    } else {
      qs.logger.log(compileUpgradeResult.compiledUpgrades.scripts.map(s => s.description.description).join('\n'));
    }
```

**Review by aeisenberg — APPROVED** [review]
My suggestion is not a blocker for going in, but I think it would make it easier to read.

File: extensions/ql-vscode/src/upgrades.ts (hunk 1 above)
Comment (alexet) [reply extensions/ql-vscode/src/upgrades.ts:157]: Is you suggestion mostly about switching the sense of the test or about the `=== undefined`.

It has to have the `=== undefined` part to get the types correct (according to the types we could have an empty `compiledUpgradeFile`). However I can flip the sense of the test if that makes more sense to you.

Comment (aeisenberg) [reply extensions/ql-vscode/src/upgrades.ts:157]: I see why typing is broken if you avoid the `undefined`. I think just adding a comment would be sufficient for me.

File: extensions/ql-vscode/src/messages.ts [hunk 2]
```diff
@@ -487,10 +492,14 @@ export interface UpgradeDescription {
   newSha: string;
 }
 
+
+export type CompiledUpgrades = MultiFileCompiledUpgrades | SingleFileCompiledUpgrade
```
Comment (aeisenberg) [inline extensions/ql-vscode/src/messages.ts:496]: And looking at this again, it's not easily clear that these two types differ in only two properties. In this case, I wonder if using proper inheritance would make things clearer. So, something like this:

```js
export interface CompiledUpgrades {
  // add all common properties here
}

interface MultiFileCompiledUpgrades extends CompiledUpgrades {
  newDbscheme: string;
  scripts: CompiledUpgradeScript[];
  compiledUpgradeFile?: never;
}

interface SingleFileCompiledUpgrade extends CompiledUpgrades {
  descriptions: UpgradeDescription[];
  compiledUpgradeFile: string;
}
```

Also. note that I don't think you need to export these types.

**Comment by aeisenberg** [comment]
Converting this to a draft. Please mark as "ready for review" when you want me to look at it.

**Comment by alexet** [comment]
I think this is ready at the moment.

**Review by aeisenberg — COMMENTED** [review]
Looks good. Possible small improvement below.

**Comment by aeisenberg** [comment]
I've tested it with both an old codeql cli and a new one. Behaviour works as expected. :shipit:

File: extensions/ql-vscode/src/pure/messages.ts [hunk 3]
```diff
@@ -401,6 +401,11 @@ export interface CompileUpgradeParams {
    * A directory to store parts of the compiled upgrade
    */
   upgradeTempDir: string;
+  /** 
+   * Enable single file upgrades, set to true to allow
+   * using single file upgrades.
+   */
+  singleFileUpgrades?: boolean;
```
[the remark this answers is not in the document]
Comment (aeisenberg) [reply extensions/ql-vscode/src/pure/messages.ts]: Made this change.

**Comment by aeisenberg** [comment]
After this build passes, I will merge.

**Pull request merged**